### ISSUE TITLE
Localisation: Danish

### ISSUE THREAD

**Issue by PeterOeClausen** [OptiKey/OptiKey#273]
I am currently working on the Danish translation.

So far I have created an account on [Transifex](www.transifex.com).
I have submitted a join request for the Danish (da) language on the OptiKey v2 project.

Approved by OptiKey.

Now translating on [Transifex](www.transifex.com).

**Comment by PeterOeClausen** [comment]
Done translating into Danish on [Transifex](www.transifex.com).

According to step: [Translating the strings](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md#translating-the-strings) which says:

> After completing the translation in Transifex, you should notify us as soon as possible so the translation you made can be checked and have it's .resx file generated as you will need this in order to implement the new locale inside OptiKey. Having acquired your language .resx file it's time to start adding the new Locale inside OptiKey.

I should now notify OptiKey (this is done with this message), I should now acquire my language .resx file and begin the next step.

Next step: [Integrate the translation into OptiKey](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md#integrate-the-translation-into-optikey).

**Comment by PeterOeClausen** [comment]
I am currently looking for a .dic file that is specified as:
> The .dic file which is the dictionary that will be used for word prediction.

This description was found in step: [Adding the translation specific files into OptiKey](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md#adding-the-translation-specific-files-into-optikey)

Still looking for it...

**Comment by JuliusSweetland** [comment]
@PeterOeClausen I'll grab the resx file for you when we get a bit further along. The dictionary that you need is a simple word list with roughly 100k words, one per line. If you can't find a good candidate let me know with a direct message and I'll see if I can generate one for you.

**Comment by PeterOeClausen** [comment]
The only Danish .dic file I could find was already in the project, in folder: OptiKey/src/JuliusSweetland.OptiKey/Dictionaries/Candidates/

Also seen [here on GitHub](https://github.com/OptiKey/OptiKey/tree/master/src/JuliusSweetland.OptiKey/Dictionaries/Candidates).

@JuliusSweetland if you have or can generate a better one, or one you like more, then I can use that. If not, then I'll try with the one in the "Candidate" folder specified above.

**Comment by JuliusSweetland** [comment]
@PeterOeClausen I'm only on mobile do I can't see how good that dictionary file is. How many entries are there and how good does it look to you?

**Comment by PeterOeClausen** [comment]
@JuliusSweetland It has 69861 entries (so lower than 100k).

We have 3 more letters in the Danish alphabet (here written in format: UppercaseLowercase): Ææ, Øø and Åå, which in other languages are written:
æ = ae,
ø = oe,
å = aa.

I see words with letters 'ø' and 'å', but no words with 'æ'. Which is a common letter in Danish. Words with 'æ' are instead written "ae".

So it is not perfect. 

However it would be usable.

**Comment by JuliusSweetland** [comment]
@PeterOeClausen ligatures and other special characters in each language present interesting problems. In the case of ligatures like æ I decided in other languages to never use the ligature and instead use the flattened letters, i.e. ae.

For ø and å I need your advice. Are these commonly used in modern Danish? Would it be better to have all dictionary entries both ways, i.e. an entry with the special character (e.g.ø) and another with the equivalent characters (e.g. oe)?

**Comment by ysangkok** [comment]
[Ø is used in Danish, Faroese, Norwegian, and Southern Sami. Ø is also used in the orthographies of several languages of Africa, such as Lendu, spoken in the Democratic Republic of the Congo, and Koonzime, spoken in Cameroon.](https://en.wikipedia.org/wiki/Ø)

According to [this page](http://spjdrpedia.dk/wiki/Frekvensanalyse#Bogstavhyppighed), ø is the third-least frequent Danish letter, only more popular than c or y.

Since ø is on the Danish standard layout, I think most users will expect it to be there.

I think you can treat ø exactly like the German ö.

[Though the Danish, Norwegian, Swedish and Walloon character Å (å) is derived from an A with a ring, it is considered a distinct letter in those languages.](https://en.wikipedia.org/wiki/Ring_(diacritic))

**Comment by JuliusSweetland** [comment]
That's really helpful and suggests that we need both ø and å on the
keyboard and in the dictionary words. This page
https://en.m.wikipedia.org/wiki/Danish_and_Norwegian_alphabet actually
suggests that we need to include the ligature æ as well, as it is
considered a letter in the alphabet of 29 letters.

That page also suggests that we need an acute accent on the keyboard, which
can be applied to any vowel (does this include ø and å?) to indicate
stress, but is most common on e.

As such we need a dictionary that only uses æ, ø and å, and not versions of
the same words with the substitutions æ = ae, ø = oe, å = aa. Does that
sound correct?

**Comment by ysangkok** [comment]
It does sound correct, but if screen space is too sparse, I think users can quickly get used to writing the substitutions. I never saw the ring being applied as a dead-key, so that means å is usually input differently than é. For example, I use the "[English (US, international with dead keys)](https://upload.wikimedia.org/wikipedia/commons/2/22/KB_US-International.svg)" layout on X.org. I have å on altgr-w and ø on altgr-l and æ on altgr-z. But é would be `'` and then e.

**Comment by JuliusSweetland** [comment]
Perfect. Space shouldn't be a problem. Russian had something like 36
characters!

**Comment by PeterOeClausen** [comment]
@JuliusSweetland 'æ', 'ø' and 'å' is very commonly used in Danish.
Examples:
"mænd" means "men",
"mælk" means "milk",
"Brændstof" means "fuel".

"brød" means "bread",
"børste" means "brush",
"første" means "first".

"få" means "to get",
"forstå" means "understand",
"måde" means "in a specific way".

These are examples of everyday words we use in Danish which are always spelled with their letters 'æ', 'ø' and 'å'. We do have words which are spelled "the old way" as "Aarhus" which is a city in Denmark, but is considered "old fashioned".

Letters on our keyboards always include the letters 'æ', 'ø' and 'å' as seen on the right of the keyboard: [here](http://fontmeme.com/images/danish-keyboard-550x183.png)


We do sometimes use the accent '´' on letters like e to make it 'é'. We do that by first pressing the '´'-key and then pressing the 'e' key on the keyboard. As such, 'é' does not have a keyboard key for itself.

On the iPhone keyboard we also have æ, ø and å keys on the right, and to produce "é", we hold in "e", a small menu comes up, and we can choose "é".

Here are some pictures of the Danish keyboard for guidance:
On iPhone:
![fil_008](https://cloud.githubusercontent.com/assets/6975621/17772036/d489da56-6546-11e6-9af4-8667a37f87d3.png)

A regular Danish laptop:
![fil_000](https://cloud.githubusercontent.com/assets/6975621/17772045/e335f580-6546-11e6-8d25-acdc14c597a3.jpeg)

How e with accent: 'é' is made on the iPhone keyboard:
![fil_006](https://cloud.githubusercontent.com/assets/6975621/17772079/022d8610-6547-11e6-9347-40c67dbdd8bc.png)

æ, ø, and å can be made the same way with an English (US) keyboard on the iPhone:
![fil_001](https://cloud.githubusercontent.com/assets/6975621/17772153/744911b0-6547-11e6-84af-f69c8d377ca5.png)

A keyboard and/or dictionary without æ, ø and å would therefore not be optimal for a Danish user.

Hope this helps :-)

**Comment by JuliusSweetland** [comment]
Really good info, thanks. Why does your laptop keyboard include letters
with an umlaut/double dot, as well as a dedicated umlaut/double dot "dead"
key?

**Comment by PeterOeClausen** [comment]
I think it is because the Keyboard layout is used in Scandinavia. 

The Swedish alphabet also has "ö" and "ä" as seen here: https://en.wikipedia.org/wiki/Swedish_alphabet

The Norwegians has the same alphabet as the Danes: https://en.wikipedia.org/wiki/Norwegian_orthography

This explains why I have both "ØÖÆ" written on my "Æ" key, and "ÆÄØ" written on my "Ø" key.
Why both Æ and Ø are on these keys, I don't know, I think they might be switched around on a Norwegian keyboard...?
When using a Danish keyboard (in windows settings), I write "Æ" with the "ØÖÆ" key and "Ø" with the "ÆÄØ" key. 
- Use the Danish iPhone keyboard when you need to map out the Danish keyboard (Æ, on the left, Ø at the right of it, and Å on top of Ø).
I have also created this picture, and colored out the letters that is not registered using the key (leaving the letter that is registered when pressing the key):
![fil_005 edited](https://cloud.githubusercontent.com/assets/6975621/17774305/5606a6fc-6553-11e6-8338-d2b98cc88016.jpg)

I rarely use the "^¨~" key, mostly (if not only) when I am writing to a German or Swedish friend in their native languages. But to produce ö and ä, I need to use the "^¨~" dead key. This explains the need for that key.

**Comment by JuliusSweetland** [comment]
Perfect. We need a 30 character keyboard then; the 29 canonical letters
(A-Z and ÆÄØ) plus an acute accent key. I'm on mobile until next week when
I can help design this and also generate a dictionary using the Firefox
spell check dictionary ordered by the most common words used in Danish
films and tv shows.

If you'd like to continue with the other localisation steps let me know and
I'll create you a Danish branch to fork and work against.

**Comment by ysangkok** [comment]
Note that Ä != Å

**Comment by JuliusSweetland** [comment]
Good point.That was just a bad copy/paste from me on mobile.

**Comment by PeterOeClausen** [comment]
@JuliusSweetland Sounds great with the 29 letters (A-Z and ÆØÅ) and the acute accent key.

We did find some Danish word lists, but they weren't great (a lot of non-Danish words and weird characters was in the mix, and we would rather not use that), we would much like to see the generated dictionary using Firefox spell checking dictionary :-)

Can you create the Danish branch to fork and work against, then I will see what I can do for the further development Localisation of OptiKey. :-)

**Comment by JuliusSweetland** [comment]
New Danish branch is here: https://github.com/OptiKey/OptiKey/tree/Danish

I'll generate the new dictionary on Sunday when I'll next have access to my
laptop and my LinqPad scripts.

**Comment by PeterOeClausen** [comment]
@JuliusSweetland Sounds great Julius, and thank you :-)

**Comment by JuliusSweetland** [comment]
@PeterOeClausen I'm struggling to get a Danish candidate dictionary. What has worked for me before is to use an OpenOffice dictionary and filter it by tv/movie substitle data to extract the most popular X words. The usual steps are to download a Danish oxt file, from which you extract the culture code .dic and .aff files (e.g. this one http://extensions.libreoffice.org/extension-center/stavekontrolden-danish-dictionary) then use the HunSpell unmunch tool to extract a full dictionary of all words. This should be achieved with this command: unmunch da-DK.dic da-DK.aff >> el-GR.unmunched

Unfortunately this is resulting in a mixed up .unmunched file. Do you have any decent candidated dictionaries? The bigger the better as I'll filter them down.

**Comment by PeterOeClausen** [comment]
@JuliusSweetland Ok.
I have found this file: 
[danish (unedited).xlsx](https://github.com/OptiKey/OptiKey/files/430180/danish.unedited.xlsx)
It is from this webpage: http://www.101languages.net/danish/most-common-danish-words/

The data is from "created using public/free subtitles, from opensubtitles in particular."
and is under the license "Creative Commons": http://creativecommons.org/licenses/by-sa/3.0/

From this file my plan to generate a .dic file was:
-Delete notes.
-Take the first 30000 words and extract them. (If we take much more than 30000 words, fewer Danish words are in the mix, and weird characters start to become more frequent).
This is the file I got in this step: 
[danish 30000 most used words (unsorted).xlsx](https://github.com/OptiKey/OptiKey/files/430187/danish.30000.most.used.words.unsorted.xlsx)

-Then I sorted it alphabetically (from A to Å): 
[danish 30000 most used words (alphabetically).xlsx](https://github.com/OptiKey/OptiKey/files/430192/danish.30000.most.used.words.alphabetically.xlsx)

-My plan was then to remove some of the characters that we do not have in the Danish language, and some words we don't have.
-Then save it as a .txt file using excel.
[danish 30000 most used words (alphabetically).txt](https://github.com/OptiKey/OptiKey/files/430221/danish.30000.most.used.words.alphabetically.txt)

-Then save it as a .dic file using Sublime Text 3.
(Github does not support this file type, so I can send it to you by email instead?).

What is your opinion on this?
-Can it be done this way?
-Do you think 30000 words is too few?
It might still contain words with weird characters, but maybe we can sort them out somehow.


The file is not perfect as you can see. However we could start with this file, and if we find a better one, we use that instead. John and I have contacted some people, who might have a better file to use. We are awaiting an answer (if we get an answer).
However this file would be a great start.

**Comment by ysangkok** [comment]
A better source would be http://da.speling.org/ , this word list is widely used in e.g. LibreOffice.

**Comment by PeterOeClausen** [comment]
@ysangkok That is perfect! Nice work!

Found this on the site @ysangkok suggested:
[dsdo-1.6.36.zip](https://github.com/OptiKey/OptiKey/files/430272/dsdo-1.6.36.zip)

The file "words-da" is perfect :-)

**Comment by JuliusSweetland** [comment]
@ysangkok @PeterOeClausen dsdo-1.6.36.zip looks good to me. I've merged in the candidate dictionary that was in the Candidate folder in the project and cleaned the combined list up a bit. It's committed on the Danish branch - let me know how it looks to you.

**Comment by PeterOeClausen** [comment]
@JuliusSweetland Unfortunately, the combined list still lacks words with the letter "Æ".
I don't know if we came to an agreement when we talked about it earlier, but "Æ" in Danish is a letter just like "A", "B", "C" and "Ø" and "Å".
Therefore the dictionary should also contain word with the letter "Æ" and not it's translation "ae".

I think the best solution would be to use the file named "words-da" from dsdo-1.6.36.zip, and make that into the .dic file. Maybe we should cut it down because it is quite long, but it is the best Danish word list we have found so far.

And we should not use the candidate file in the project (because it does not have the letter Æ), we should only use the "words-da" file :-)

**Comment by ysangkok** [comment]
Why are you referring to version 1.6.36 all the time, even though it is not the newest one?

The newest one right now: http://da.speling.org/filer/dsdo-1.7.42.zip

**Comment by PeterOeClausen** [comment]
@ysangkok I did not find the version you are now referring to, but now that you have linked it directly, we might as well use that one :+1:

**Comment by JuliusSweetland** [comment]
@PeterOeClausen I've updated the DanishDenmark.dic file on the Danish branch from the linked dsdo-1.7.42.zip file. The previous version and the current version both contain words with the letters "Æ", "Ø" and "Å". If you aren't seeing these characters try reopening the file with Encoding="UTF-8".

**Comment by PeterOeClausen** [comment]
@JuliusSweetland Now it's in the dictionary folder, which it wasn't when I checked before (maybe I forgot to pull the latest changes, I'm not sure).
Now I have pulled the latest changes, and the DanishDenmark.dic file looks great. (It is however not sorted alphabetically from A to Å), but I don't know if this matters for the application?

Also it's 337514 words long, is that alright compared to the 100k you talked about earlier? Or does it need to be made smaller?

Great work so far.
Shall I begin with the next step on Localizing OptiKey into Danish? Or are there more things that needs to be done before I continue?

**Comment by JuliusSweetland** [comment]
Alphabetical order is not important. You can continue. The size is a little
large, but we can test and filter it later.

**Comment by PeterOeClausen** [comment]
@JuliusSweetland Sounds like a great plan.
I'll continue :-)

**Comment by PeterOeClausen** [comment]
@JuliusSweetland I have followed [the guide](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md) all the way down to the step called ["Add a keyboard layout"](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md#add-a-keyboard-layout).

The only places I had to not follow the guide was when described:
> Now the modification and regeneration of Resources.Designer.cs file will be done automatically by Visual Studio itself after it detects a changed Resources.resx file; thus at this point just rebuild the application -- which should happen without any problems.

I had to write them in manually like this (because they were not automatically done when rebuilding, building or running the application):
![image](https://cloud.githubusercontent.com/assets/6975621/17934725/43fbdf5c-6a19-11e6-8dee-a97eef950bb3.png)


Also in step: ["Add Language Key in OptiKey KeyValues"](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md#add-language-key-in-optikey-keyvalues)
I added the following:
![image](https://cloud.githubusercontent.com/assets/6975621/17934876/caebb91a-6a19-11e6-90ef-dc8490dad5de.png)

I was not sure if Æ, Ø and Å are "non-Latin characters", but I have inserted them as I think they are not.
(If you know otherwise, feel free to correct me, and I'll go and change it to "defaultList").


I have tried to push my work to the Danish branch, but I get a 403: Forbidden. It looks like I am on the Danish branch as well. I don't know if I am allowed to push to the Danish branch?

Here is a Screen snippet of what I have tried so far:
![image](https://cloud.githubusercontent.com/assets/6975621/17935873/274e675e-6a1d-11e6-986e-f7e19ac7560c.png)

By the way, I am using "Git Shell" from the ["GitHub for Desktop"](https://desktop.github.com/), on Windows 10.

I did also try to commit/pull/push using the GUI, but it got stuck, so I tried the Git Shell instead (Not sure if this can cause a problem).

I don't have anymore time to work on OptiKey today. But I can continue tomorrow. I can try looking into fixing this Git stuff if I should have permission to push? Also I can look at the ["add a keyboard layout" step](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md#add-a-keyboard-layout).

Best, Peter.

**Comment by JuliusSweetland** [comment]
Hi Peter,

Everything looks fine. I'll have a play with the resources this weekend to
see why they're not working for you. There's a bit of manual stuff I need
to do with resources anyway.

The multi key section character string looks perfect.

I'm on mobile, but it looks like you're trying to work against my version
of the Danish branch, which you can't edit unless you're a member of the
OptiKey organisation (a GitHub concept). You need to fork the OptiKey
branch so you have your own copy of the repo. You then make your changes
against that, commit and push them to your repo, and then submit a pull
request to me (well, the OptiKey organisation). This gives me a chance to
look over potential charges, test, suggest changes, etc, before I accept
the PR and your changes are merged in. After that I'd manually prepare a
release candidate to test.

**Comment by PeterOeClausen** [comment]
Hi @JuliusSweetland 

Ok. I have now forked the Optikey Danish branch repo.
Cloned the project locally.
Made the changes: All steps of the guide until (not including) ['Add a keyboard layout'](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md#add-a-keyboard-layout).
- If you want, I can continue with the 'Add a keyboard layout' step on the forked and cloned repository.

I have pushed my changes and submitted a pull request.

I still get the
![image](https://cloud.githubusercontent.com/assets/6975621/17971970/49836fb6-6add-11e6-8967-60d7e8fa0be7.png)
Warning message when pushing.

When running the project, it looks like it runs great (no error messages or crashes). Still need the Danish Keyboard though.

Also the text strings for keys like "Mouse" (English) should be "Mus" (Danish), but they are still in English (I think it is the same on the German language setting).

The Danish dictionary works when typing.

The Danish language setting can take a bit longer to load compared to other language settings.

Hope this helps.
Tell me if I need to do some changes to the forked repo, and if I should continue with the 'Add a keyboard layout' step.

Best, Peter.

**Comment by JuliusSweetland** [comment]
Great work. You can continue to create a keyboard layout if you'd like.
You'll need 4 keys including the new accent dead key.

I think you can ignore the resource line ending warning.

I need to look at the missing resource strings - I'll look on the weekend
at everything.

Great work so far. Feel free to create a PR when you're happy. It doesn't
need to be absolutely complete as your changes will initially be merged
onto a feature branch on my side, not master.

**Comment by JuliusSweetland** [comment]
@PeterOeClausen Hi Peter. I've just tested the German strings and I couldn't reproduce the problem with "Mouse" not showing correctly in German:

![image](https://cloud.githubusercontent.com/assets/4722561/18033727/11088812-6d24-11e6-937f-d29c239cff89.png)

When you're ready can you create a PR and I'll look into the Resource stuff.

I was also thinking about your problem with Resources.Designer.cs not automatically regenerating - I think you have to rebuild the project.

**Comment by PeterOeClausen** [comment]
I have tried downloading the newest version (2.3.5.0), from the OptiKey.org website (this afternoon).
Here, when switching the Language setting to German. The keys strings doesn't change:
![optikeyorginstallation](https://cloud.githubusercontent.com/assets/6975621/18170691/7f0f5fbe-705f-11e6-8456-c79f0ae4591c.png)

Also when running the project I currently have on the Danish branch (version 2.3.6.0) and switching to German:
![optikeydanishbranch](https://cloud.githubusercontent.com/assets/6975621/18170859/210093ec-7060-11e6-985a-ae05b816f184.png)

To change the language setting, I go to the main menu, and mouse of the "LANGUAGE" button. Then I mouse over "GERMAN (GERMANY) / DEUTSCH (DEUTSCHLAND)" until the circle is full, a short moment after it switches back to the main menu. And the text strings are not changed.

I am running OptiKey on Windows 10. I also tried the same with the OptiKey.org installation (version 2.3.5.0) on a computer running Windows 7, the results were no different.

I hope this helps.

Also, I will start working on the Danish keyboard as specified in the [Add a keyboard](https://github.com/OptiKey/OptiKey/blob/master/docs/Localize.md#add-a-keyboard-layout) step of the guide.

I hope I can still work on the Danish branch as it is now. Tell me if I need to update my branch or fork it again, before implementing the keyboard.

Best, Peter.

**Comment by JuliusSweetland** [comment]
Hi Peter. Ok, that's actually how it is supposed to be - changing the
language in the way that you are changes only the keyboard layout, not the
language used in the UI strings. The reason is that you might have the UI
language set to Russian, but want to type a few words in English, but
you're Russian so you wouldn't want the UI strings to change. If you want
to change the UI strings then you must open the management console, go to
the words tab, and change it there - you'll notice there are two settings;
keyboard language (which you have been changing), and the UI language.

Let me know if that doesn't make sense.

**Comment by PeterOeClausen** [comment]
Hi Julius.
I have now completed the Danish localisation on a forked repository located at https://github.com/PeterOeClausen/OptiKey.

I will now report on some of the changes I have made:

1) I've moved some buttons around in the language menu, so that there were room for "Danish", and so that the back button is still in the same place:
![language menu](https://cloud.githubusercontent.com/assets/6975621/18548984/f0139b2a-7b4b-11e6-96a5-c272848f942f.png)

2) I have designed and build a Danish keyboard:
![danish keyboard](https://cloud.githubusercontent.com/assets/6975621/18548994/00563bc8-7b4c-11e6-9330-80889f40395c.png)

Note that I have removed the ABC key:
![removed abc key](https://cloud.githubusercontent.com/assets/6975621/18548999/0b313ef8-7b4c-11e6-8853-fc4901a0d44f.png)
Because it was a redundant button, in that it just went to the same screen the user is already using. I found no need for it.
Also I have moved the mouse buttons a bit, so that the keyboard reminds me more of a Danish keyboard. 
This is the solution I like the most. What do you think?

3) I have designed and build a Danish conversation keyboard:
![conversation keyboard](https://cloud.githubusercontent.com/assets/6975621/18549021/307c37f8-7b4c-11e6-8a2a-75961b69cdac.png)

4) When testing out the keyboard and dictionary in relation to the accented e: é and É, the most frequent used words that uses them: én, ét and idé is suggested by the dictionary. Therefore I see no real need for an 'é' button, or an accent button.
This is illustrated here:
![e with accents](https://cloud.githubusercontent.com/assets/6975621/18549125/ca021640-7b4c-11e6-84d1-7bf35a4b8f4e.png)

5) I do get some warnings and one error when building and running the application. The application runs fine even though these errors are in the error list.

The error/warning messages:
![image](https://cloud.githubusercontent.com/assets/6975621/18549184/2cfb7cc8-7b4d-11e6-884a-879fb8a2feb6.png)

6) I have created a pull request on the OptiKey/OptiKey Danish branch with my PeterOeClausen/OptiKey master branch:
![image](https://cloud.githubusercontent.com/assets/6975621/18549329/21a4e642-7b4e-11e6-8ecc-ebaf2b44b31a.png)

Do you want me to send a pull request to the OptiKey/OptiKey master branch, or the Danish branch.
Github says my master branch and your master branch can be automatically merged:
![image](https://cloud.githubusercontent.com/assets/6975621/18549364/5606be7e-7b4e-11e6-8726-dda1efa05489.png)

I think you want me to create a pull request to the Danish branch, so that is what I submit. Please correct me if I'm wrong.

7) Also you were right. I didn't know I had to go to the management console (ALT+M) to change the button labels. It now works for me as well.

Sorry for the long message.
Best, Peter.

**Comment by JuliusSweetland** [comment]
@PeterOeClausen Hi Peter - excellent detail in your post. Really useful!

Everything you've done looks fantastic. If the users miss the accents they can always use the diacritic keyboards to find the accented letter explicitly.

What is the Object reference exception error you are getting? I'm sure it can be ignored as a false positive if it is building, but it would be interesting to check.

You are correct; I would like the pull request against the OptiKey/Danish branch - I will then check it out this weekend.

How is the performance with the large dictionary?

Great work! It's looks really, really good.

**Comment by PeterOeClausen** [comment]
@JuliusSweetland 
_"If the users miss the accents they can always use the diacritic keyboards to find the accented letter explicitly"._
I agree. And if the Danish users want it differently, we can change it later on.

_"What is the Object reference exception error you are getting? I'm sure it can be ignored as a false positive if it is building, but it would be interesting to check"._
I don't know the detail behind it. I had it in the Alpha.xaml keyboard as well earlier, but it disappeared somehow.
The blue squiggly lines under "controls:KeyboardView" in the ConversationAlpha.xaml, (shown in the image I send) with mouse over it has the same text as in the error message in the bottom: "Object reference not set to an instance of an object.".
I don't know what is causing this, but it runs flawlessly.

The performance with the large dictionary is fine. I can't see a difference from the English dictionary (no stutter of lag of any kind) - tested on a quad core laptop though. Might be different on a tablet.

Sometimes the suggestions are not quite what I want to write, but this can be optimized in different ways. The suggested words are typically shorter than the word I want to write, so I typically end up spelling the whole word before I can find it in the suggestion area. This can be optimized in different ways, for example by keeping track of the user's most used words, and suggesting these before other words. However this is much more work to implement later in the future. (Tbh. I haven't looked into how the suggestions are implemented in OptiKey yet). The keyboard is really useful nonetheless.

All in all great! I look forward to hear from you when you have checked it out :-)

Best, Peter.

**Comment by JuliusSweetland** [comment]
@PeterOeClausen Pre-release v2.3.6 including Danish (and Slovenian) is available here; https://github.com/OptiKey/OptiKey/releases - can you try it out and let me know if it works as expected?

**Comment by PeterOeClausen** [comment]
Hi @JuliusSweetland.
My boss and I have checked out the Danish version with an eyetracker.

I have found one small translation error, that I will correct on Transifex, and I will correct it in the program and create a new pull request. (I will notify when this is done).

**Other than this we would say it's ready to be released.**

My boss and I have also talked about different optimizations that could be done to OptiKey, that would probably require a lot more work, but would make the user experience better. This is a list of things we came up with:

- Optimization of the dictionary:

We found the Danish dictionary to be usable, but not perfect.
It uses the words that are closest alphabetically to the written word, and since the Danish dictionary is quite large, a lot of unwanted contractions and abbreviations clutter up the suggesting area.

Because of this, we found that we either spelled the whole word with each character, without using or even looking at the suggestions (because it would be faster to just spell out each character). However when we used the "Multi-key selection" tool, and have OptiKey guess the word we wanted to spell, it guessed right the most of the time. If the word was not right, we use the suggestions to correct it. In the latter scenario, the dictionary becomes very handy.

The first scenario could be optimized by suggesting the most frequent used words in a language first, before abbreviations and contractions.

This could otherwise be optimized by analyzing the sentence after each word is entered. So if the user wrote "I think this is very" and then entering letters "i", "m" and "p", OptiKey would suggest "important" before "import", (this is just an example, I am not sure how it is implemented in practice).

This of course is an optimization, it is not needed to use the program, but it would be nice if it was there.


- Use mouse magnifier to precise mouse clicking.

My boss and I played around with the "Magnifier" feature after selecting the "Mouse" key on the keyboard.
After fixating on a point on the screen, a square shows up in the middle of the screen with a zoomed in view of the area around the point I was looking at.
It would be really nice if the buttons and fields inside this box was clickable with left and right mouse key, so that I could use the feature to easily navigate smaller buttons on the screen.

We actually thought this was a feature that did not work properly (please tell me if it is supposed to work like this, because it did not work when we tried it out).


- Changing the speech language when changing keyboard language.

In some bilingual countries, it would be nice to easily switch between two languages, maybe just by changing the keyboard setting.
We have a Danish SAPI5 voice installed on the computer we were using, and OptiKey used this to pronounce Danish words perfectly. However when trying to pronounce English words, the Danish SAPI5 voice did not do well.


- Bug in "conversation" keyboard when resizing OptiKey.

Normally the conversation keyboard fill up the whole screen which is nice.
After adjusting the size of OptiKey in the settings, the conversation keyboard did not start full screen as it used to.
This bug can also be caused by other things on our PC, and I need to test it more to be able to pinpoint when it is happening. We just noticed this when using it.

Best, Peter.

**Comment by PeterOeClausen** [comment]
I have corrected the translation error on Transifex, downloaded the resource, changed the Resource.dk-DK.resx file, tested the changes, and submitted a pull request from my fork to the OptiKey Danish branch.

Best, Peter.

**Comment by JuliusSweetland** [comment]
@PeterOeClausen Thanks Peter. I have merged in your Resources fix and will release a new pre-release this weekend.

* Optimization of the dictionary: The size of the dictionary is always a balancing act. Too large and the suggestions when typing letter-by-letter are not great, but the multi-key selections (swiping) are generally very good. Too small and neither a great. OptiKey does keep a count of how often dictionary words are used and these words then rise to the top of the suggestions over time. Would you like to reduce the dictionary size?

* Use mouse magnifier to precise mouse clicking: The magnify feature is meant to work like this;
a) The user selects the mouse action key, e.g. right click.
b) The user fixates roughly where they want to click and a magnified view of that area is presented to the user.
c) The user the fixates on the magnified image exactly where they want to right click and the right click is performed on the corresponding point on the screen once the magnified view is closed.
d) Alternatively the user can selection outside of the magnified view, which cancels the mouse action.

* Changing the speech language when changing keyboard language: This is definitely possible, but not something I included, as I thought users would want a consistent voice throughout their experience, even if the voice mispronounced foreign languages a bit. The technical challenge with a automatically switching voices is that the user would need to have a voice installed for each language available in OptiKey, and OptiKey would need to know which voice matches which language. It's all solveable, but need some thought and discussion on whether this is feasible and useful.

* Bug in "conversation" keyboard when resizing OptiKey: How did you change the size in the settings? OptiKey doesn't have sizing settings as far as I know?

**Comment by JuliusSweetland** [comment]
@PeterOeClausen Hi Peter - can you let me know if you would like to optimise (shrink) they dictionary before Danish is ready to go? I am ready to make a new release when you are... thanks.

**Comment by PeterOeClausen** [comment]
@JuliusSweetland Hi Julius.
Sorry for answering late, I've been busy.

_"OptiKey does keep a count of how often dictionary words are used and these words then rise to the top of the suggestions over time."_
Great. This will improve the use of the Dictionary over time for the individual user.

_"can you let me know if you would like to optimize (shrink) the dictionary before Danish is ready to go?"_
I think the current dictionary size is great.
I do not believe it is necessary to spend hours going through and manually cutting it. And I have to think further if I were to cut a set of words out of it (for example words shorter than 3 characters, what would happen to multi-key function if this was done and so on...).
My boss and I think the dictionary is great at it's current state. Of cause it can be optimized, however this is not worth the time for me right now.

_"Bug in 'conversation' keyboard when resizing OptiKey: How did you change the size in the settings? OptiKey doesn't have sizing settings as far as I know?"_
and
_"Use mouse magnifier to precise mouse clicking: The magnify feature is meant to work like this; (...)"_
I have encountered bugs with these things not working properly. However the bugs happened on a computer where each installation of OptiKey was installed on top of the previous. A fresh install of OptiKey might solve these issues. I will look into these bugs further, and test whether a fresh install of OptiKey would fix these issues. However I do not think these issues has nothing to do with the new release.

In my (and my boss's) opinion, the new release is ready. 

Best, Peter.

**Comment by JuliusSweetland** [comment]
Great, thanks. I'll prepare a release soon.

**Comment by JuliusSweetland** [comment]
Released in v2.3.6
